packages/query-core/src/streamedQuery.ts

```diff
@@ -6,46 +6,63 @@ import type { QueryFunction, QueryFunctionContext, QueryKey } from './types'
  * The query will be in a 'pending' state until the first chunk of data is received, but will go to 'success' after that.
  * The query will stay in fetchStatus 'fetching' until the stream ends.
  * @param queryFn - The function that returns an AsyncIterable to stream data from.
- * @param refetchMode - Defaults to 'reset', which replaces data when a refetch happens. Set to 'append' to append new data to the existing data.
+ * @param refetchMode - Defines how re-fetches are handled.
+ * Defaults to `'reset'`, erases all data and puts the query back into `pending` state.
+ * Set to `'append'` to append new data to the existing data.
+ * Set to `'replace'` to write the data to the cache at the end of the stream.
  */
 export function streamedQuery<
   TQueryFnData = unknown,
   TQueryKey extends QueryKey = QueryKey,
 >({
   queryFn,
-  refetchMode,
+  refetchMode = 'reset',
 }: {
   queryFn: (
     context: QueryFunctionContext<TQueryKey>,
   ) => AsyncIterable<TQueryFnData> | Promise<AsyncIterable<TQueryFnData>>
-  refetchMode?: 'append' | 'reset'
+  refetchMode?: 'append' | 'reset' | 'replace'
 }): QueryFunction<Array<TQueryFnData>, TQueryKey> {
   return async (context) => {
-    if (refetchMode !== 'append') {
-      const query = context.client
-        .getQueryCache()
-        .find({ queryKey: context.queryKey, exact: true })
-      if (query && query.state.data !== undefined) {
-        query.setState({
-          status: 'pending',
-          data: undefined,
-          error: null,
-          fetchStatus: 'fetching',
-        })
-      }
+    const query = context.client
+      .getQueryCache()
+      .find({ queryKey: context.queryKey, exact: true })
+    const isRefetch = !!query && query.state.data !== undefined
+
+    if (isRefetch && refetchMode === 'reset') {
+      query.setState({
+        status: 'pending',
+        data: undefined,
+        error: null,
+        fetchStatus: 'fetching',
+      })
     }
+
+    const result: Array<TQueryFnData> = []
     const stream = await queryFn(context)
+
     for await (const chunk of stream) {
       if (context.signal.aborted) {
         break
       }
-      context.client.setQueryData<Array<TQueryFnData>>(
-        context.queryKey,
-        (prev = []) => {
-          return prev.concat(chunk)
-        },
-      )
+
+      // don't append to the cache directly when replace-refetching
+      if (!isRefetch || refetchMode !== 'replace') {
+        context.client.setQueryData<Array<TQueryFnData>>(
+          context.queryKey,
+          (prev = []) => {
+            return prev.concat([chunk])
```

Comment:
```diff
- return prev.concat(chunk)
+ return prev.concat([chunk])
```

this tiny change is hard to see, but it fixes an issue where `TQueryFnData` is in itself an array, which got flattened implicitly by `concat`

Reply:
added a test for this, too: 0de02229